src/graph/graph.c

```diff
@@ -680,10 +674,288 @@ void Graph_DeleteNode(Graph *g, Node *n) {
+
+    // Populate mask with implicit edges, take note of deleted nodes.
+    for(uint i = 0; i < node_count; i++) {
+        GrB_Index src;
+        GrB_Index dest;
+        Node *n = nodes + i;
+        bool depleted = false;
+        NodeID ID = ENTITY_GET_ID(n);
+
+        // Outgoing edges.
+        GxB_MatrixTupleIter_iterate_column(adj_iter, ID);
+        while(true) {
+            GxB_MatrixTupleIter_next(adj_iter, &dest, NULL, &depleted);
+            if(depleted) break;
+            GrB_Matrix_setElement_BOOL(Mask, true, dest, ID);
+        }
+
+        depleted = false;
+
+        // Incoming edges.
+        GxB_MatrixTupleIter_iterate_column(tadj_iter, ID);
+        while(true) {
+            GxB_MatrixTupleIter_next(tadj_iter, &src, NULL, &depleted);
+            if(depleted) break;
+            GrB_Matrix_setElement_BOOL(Mask, true, ID, src);
+        }
+
+        GrB_Matrix_setElement_BOOL(Nodes, true, ID, ID);
+    }
+
+    GrB_Index nvals;
+    GrB_Matrix_nvals(&nvals, Mask);
+    *edge_deleted += nvals;
+
+    // Clear updated output matrix before assignment.
+    GrB_Descriptor_set(desc, GrB_OUTP, GrB_REPLACE);
+
+    // Free and remove implicit edges from relation matrices.
+    int relation_count = Graph_RelationTypeCount(g);
+    for(int i = 0; i < relation_count; i++) {
+        GrB_Matrix R = _Graph_GetRelationMap(g, i);
+        
+        // Reset mask descriptor.
+        GrB_Descriptor_set(desc, GrB_MASK, GxB_DEFAULT);
+
+        /* Isolate implicit edges.
+         * A will contain all implicitly deleted edges from R. */
+        GrB_Matrix_apply(A, Mask, NULL, GrB_IDENTITY_UINT64, R, desc);
+        
+        /* Free each multi edge array entry in A
+         * Call _select_op_free_edge on each entry of A. */
+        GxB_select(A, GrB_NULL, GrB_NULL, selectop, A, g, GrB_NULL);
+
+        // Clear both relation matrix and its coresponding relation mapping matrix.
+        GrB_Descriptor_set(desc, GrB_MASK, GrB_SCMP);
+
+        // Remove every entry of R marked by Mask.
+        GrB_Matrix_apply(R, Mask, NULL, GrB_IDENTITY_UINT64, R, desc);
+
+        R = Graph_GetRelationMatrix(g, i);
+        // Remove every entry of R marked by Mask.
+        GrB_Matrix_apply(R, Mask, NULL, GrB_IDENTITY_UINT64, R, desc);
+    }
+
+    /* Descriptor:
+     * GrB_MASK, GrB_SCMP
+     * GrB_OUTP, GrB_REPLACE */
+    GrB_Descriptor_set(desc, GrB_MASK, GrB_SCMP);
+
+    // Update Adjacency and transposed adjacency matrices.
+    GrB_Matrix_apply(adj, Mask, NULL, GrB_IDENTITY_UINT64, adj, desc);
+    GrB_Matrix_apply(tadj, Mask, NULL, GrB_IDENTITY_UINT64, tadj, desc);
+
+    /* Delete nodes
+     * All nodes marked for deleteion are detected, no incoming / outgoing edges. */
+    int node_type_count = Graph_LabelTypeCount(g);
+    for(int i = 0; i < node_type_count; i++) {
+        GrB_Matrix L = Graph_GetLabelMatrix(g, i);
+        GrB_Matrix_apply(L, Nodes, NULL, GrB_IDENTITY_BOOL, L, desc);
```

Comment:
I wonder if there are faster calls for reducing the label matrices? I don't actually know, but there are a lot of options to accomplish this particular deletion.